src/t_hash.c

```diff
@@ -2030,3 +2085,116 @@ void hrandfieldCommand(client *c) {
+    void **entries;          /* array of expired entries to replicate and free later */
+} expiryContext;
+
+/* Add an entry to the expiryContext list of processed entries. */
+static void ctxAddEntry(expiryContext *ctx, void *e) {
+    ctx->entries[ctx->n_entries++] = e;
+}
+
+/* Callback for popping expired entries from the volatile set.
+ * Deletes the entry from the hash table and tracks it in the expiry context.
+ * Returns 1 if deleted, 0 if nothing to do.
+ */
+static int expireEntry(void *entry, void *c) {
+    expiryContext *ctx = c;
+    robj *o = ctx->key;
+    serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
+
+    hashtable *ht = o->ptr;
+    void *entry_ptr = NULL;
+    int deleted = hashtablePop(ht, entry, &entry_ptr);
+
+    if (deleted) {
+        debugLogField(o, entry_ptr);
+        ctxAddEntry(ctx, entry_ptr);
+        server.stat_expiredfields++;
+        return 1;
+    }
+    return 0;
+}
+
+/* Extract expired entries from a hash object's volatile set.
+ * Returns number of expired entries, populates `out_entries`. */
+size_t hashTypeExtractExpiredEntries(robj *o, mstime_t now, unsigned long max_entries, void **out_entries) {
+    serverAssert(max_entries > 0 && max_entries <= 1024);
+
+    /* skip TTL checks temporarily (to allow hashtable lookup) */
+    hashTypeIgnoreTTL(o, 1);
+
+    vset *vset = hashTypeGetVolatileSet(o);
+    if (vsetIsEmpty(vset)) {
+        hashTypeIgnoreTTL(o, 0);
+        return 0;
+    }
+
+    expiryContext ctx = {.key = o, .entries = out_entries, .n_entries = 0};
+    size_t expired = vsetPopExpired(vset, entryGetExpiry, expireEntry, now, max_entries, &ctx);
+    serverAssert(ctx.n_entries <= max_entries);
+    hashTypeIgnoreTTL(o, 0);
+    return expired;
+}
+
+/* Hashtable scan callback for hash datatype */
+static void activeDefragHashTypeEntry(void *privdata, void *element_ref) {
+    entry **entry_ref = (entry **)element_ref;
+    entry *old_entry = *entry_ref, *new_entry = NULL;
+    long long old_expiry = entryGetExpiry(old_entry);
+
+    new_entry = entryDefrag(*entry_ref, activeDefragAlloc, activeDefragSds);
+    if (new_entry) {
+        /* In case the entry is tracked we need to update it in the volatile set */
+        if (entryHasExpiry(new_entry)) {
+            // We don't need to pass the db because db-level tracking isn't going to change for this update.
+            hashTypeTrackUpdateEntry(privdata, old_entry, new_entry, old_expiry, entryGetExpiry(new_entry));
+        }
+        *entry_ref = new_entry;
+    }
+}
+
+size_t hashTypeScanDefrag(robj *ob, size_t cursor, int (*defragRaxNodefn)(raxNode **)) {
+    serverAssert(ob->type == OBJ_HASH && ob->encoding == OBJ_ENCODING_HASHTABLE);
+    static struct volatileSetCursor {
+        size_t cursor;
+        bool is_vsetDefrag;
+    } volaSetIter;
+    static struct volatileSetCursor *vset_cursor = NULL;
+
+    vset_cursor = (struct volatileSetCursor *) cursor;
+
+    if (!vset_cursor) {
+        // New object scan
```

Comment:
you should also include hashtableDefragTables when starting to scan a new object